### DISCUSSION TITLE
JavaScript 실행되면 html 파싱 중단되는 이유에 대해

### DISCUSSION THREAD

**Discussion by skyhanull** [wanted-9th-3team/frontend-study#29] [프론트 기술 질문 공간]
JavaScript 실행되면 html 파싱 중단되는 이유에 대해 설명해주세요

**Comment by WooGie911** [comment]
브라우저 렌더링과 관련되어 있습니다.

JavaScript 실행되면 html 파싱 중단되는 이유를 간단하게 말하면 
JS 엔진과 렌더링 엔진이 직렬적으로 실행되기 때문입니다.

렌더링 과정 중에 JS 파싱과 실행이 종료되면 그 때 다시 제어 권한이 렌더링 엔진으로 넘어가 중단된 시점부터 다시 시작됩니다.

이러한 HTML 파싱 블로킹현상을 근본적으로 해결하기 위해 HTML5부터 async와 defer 추가되었으며

async와 defer는 HTML파싱과 외부 자바스크립트 파일의 로드가 비동기적으로 동시에 진행되도록 만들어줍니다.

**Comment by michoball** [comment]
브라우저 엔진은  렌더링 엔진과 js 엔진이 서로 동기적으로 실행되기 때문입니다. 
html문서를 위에서 아래로 읽으면서 렌더링을 합니다. 
그러다가 js script tag가 들어간 곳에 도달하게 되면 그때 렌더링을 멈추고 js파일을 불러와 실행하게됩니다. 
이때 렌더링 엔진은 작동 권한이 없고 js 엔진만 작동 권한이 있기 때문에 html 파싱이 중단 되는 것입니다. 

그래서 예전에는 js 관련 script tag를 body의 맨 밑에 두는 것이 일반적이었지만 최근에 비동기적으로 js를 받아오는 방법이 생겨서 이제는 head 쪽에 async 나 defer 옵션을 가지는 script tag를 두고 있습니다. 

개인적으로 defer 옵션이 페이지 렌더링에 가장 효율적이라고 생각해서 간단한 js 토이 프로젝트를 진행할 당시에는 defer 옵션만을 사용해서 구현을 했었습니다.

**Comment by bona373737** [comment]
브라우저는 서버로부터 전달받은 html,css,js파일을 가지고 화면을 그려내는 데, 
html,css파일은 렌더링 엔진에서, js파일은 자바스크립트엔진에서 해석한다. 
이때 렌더링엔진과 자바스크립트엔진은 직렬적으로 동작하기때문에 html 중간에 script태그를 만나면 자바스크립트 엔진으로 제어권이 넘어간다. 

DOM이 완성되지 않은 상태에서 자바스크립트가 DOM을 조작하면 에러가 발생할 수 있기때문에
body 요소의 가장 아래에 자바스크립트를 위치시키거나, defer속성을 사용하여 자바스크립트파일을 비동기적으로 로드하되, 실행은 DOM생성이 완료된 이후 진행되도록 하는 것을 권장한다.
